### PR TITLE
bib/internal/setup: Tag validate early

### PR DESCRIPTION
**Pull request by bearrito**
This is my first attempted commit to this project.

I'm still learning the repo structure and dev workflow. Few questions on testing.

1. In this commit I'm relying on having images locally. I didn't see anything similar so I assume we want to avoid this/tests will fail. Is there an alternative?
2. If I have to rely on testing by hand, how can I use local changes? I'm thinking I just  https://github.com/osbuild/bootc-image-builder?tab=readme-ov-file# building and then can  follow examples?

Closes: https://github.com/osbuild/bootc-image-builder/issues/380

### CODE REVIEW

**Comment by bearrito** [comment]
Working, testing by hand

```
bstrausser@LOCLAP680:~/Git/bootc-image-builder$ sudo podman run     --rm     -it     --privileged     --pull=newer     --security-opt label=type:unconfined_t     -v $(pwd)/config.toml:/config.toml     -v $(pwd)/output:/output     localhost/bootc-image-builder:tag-validate   --type qcow2 quay.io/centos/centos:stream9
WARNING: The same type, major and minor should not be used for multiple devices.
... elided... ( I don't know enough to know what the warnings mean)
WARNING: The same type, major and minor should not be used for multiple devices.
Generating manifest manifest-qcow2.json
Error: cannot build manifest: image quay.io/centos/centos:stream9 is not a bootc image
2024/06/15 21:02:07 error: cannot build manifest: image quay.io/centos/centos:stream9 is not a bootc image
```

File: bib/internal/setup/setup_test.go [hunk 1]
```diff
@@ -61,3 +61,13 @@ func TestValidateCanRunTargetArchUnexpectedOutput(t *testing.T) {
 	err := setup.ValidateCanRunTargetArch("fakearch")
 	assert.ErrorContains(t, err, `internal error: unexpected output`)
 }
+
+func TestValidateTags(t *testing.T) {
```
Comment (mvo5) [inline bib/internal/setup/setup_test.go:65]: At this level I would prefer if we would not call out to the network. We could generate a local container with the right locals via `podman build` or we just fake the podman binary in the test and also add an intergration test to the integration test suite. I lean towards the later (as the integration test is needed anyway and it is a bit easier with the existing code) - but I have no super strong opinion here. I attached to git commits that I can push here (if you want) or that you could `git am` into your tree with what I have in mind for the testing of this feature:

[0001-setup-tweak-TestValidateTags-to-fake-the-podman-bina.patch.txt](https://github.com/user-attachments/files/15882957/0001-setup-tweak-TestValidateTags-to-fake-the-podman-bina.patch.txt)
[0002-test-add-new-integration-test-for-ValidateHasContain.patch.txt](https://github.com/user-attachments/files/15882958/0002-test-add-new-integration-test-for-ValidateHasContain.patch.txt)

File: bib/internal/setup/setup.go [hunk 2]
```diff
@@ -148,3 +149,17 @@ func validateCanRunTargetArch(targetArch string) error {
 
 	return nil
 }
+
+func ValidateHasContainerTags(imgref string) error {
+	output, err := exec.Command("podman", "image", "inspect", imgref, "--format", "'{{.Labels}}'").Output()
```
Comment (mvo5) [inline bib/internal/setup/setup.go:154]: (super nitpick) I don't think the `'` around `{{.Labels}}` here is needed as this is not run in the shell but via execve() so no need to quote (but it does also not hurt).

**Review by mvo5 — COMMENTED** [review]
Thank you (and sorry for the somewhat slow reply, we are usually quicker with reviews)!

This looks very nice - I have some suggestions inline about the testing, it includes two potential git commits that could be used here to. Once the testing is sorted this looks good to go to me (but an additonal pair of eyes on this from e.g. @achilleas-k would be nice)

File: bib/internal/setup/setup_test.go (hunk 1 above)
Comment (bearrito) [reply bib/internal/setup/setup_test.go:65]: I don't like having a non-automated test, so I'm for whatever gets automation. I'll look at and apply those patches.

**Comment by bearrito** [comment]
@mvo5 I incorporated your patches.

I had an issue with  `test_manifest.py`. 5 of the tests were failing due to arch issues.  Maybe this is environmental on my end?
Example: 
```
Trying to pull quay.io/centos-bootc/fedora-bootc:eln...
Error: error creating build container: choosing an image from manifest list docker://quay.io/centos-bootc/fedora-bootc:eln: no image found in image index for architecture x86_64, variant "", OS linux
FAILED

================================================================= FAILURES ==================================================================
_______________________________________ test_manifest_disksize[quay.io/centos-bootc/fedora-bootc:eln] __
```

The specific tag tests were passing though. 

```
sudo -E /home/barrett/.local/bin/pytest -s -v test/test_manifest.py::test_manifest_checks_build_container_is_bootc

=========================================================== slowest 10 durations ============================================================
2.63s call     test/test_manifest.py::test_manifest_checks_build_container_is_bootc[quay.io/centos-bootc/centos-bootc:stream9-False-None]
1.53s call     test/test_manifest.py::test_manifest_checks_build_container_is_bootc[quay.io/centos/centos:stream9-True-image quay.io/centos/centos:stream9 is not a bootc image]
1.43s setup    test/test_manifest.py::test_manifest_checks_build_container_is_bootc[quay.io/centos/centos:stream9-True-image quay.io/centos/centos:stream9 is not a bootc image]
0.00s setup    test/test_manifest.py::test_manifest_checks_build_container_is_bootc[quay.io/centos-bootc/centos-bootc:stream9-False-None]
0.00s teardown test/test_manifest.py::test_manifest_checks_build_container_is_bootc[quay.io/centos-bootc/centos-bootc:stream9-False-None]
0.00s teardown test/test_manifest.py::test_manifest_checks_build_container_is_bootc[quay.io/centos/centos:stream9-True-image quay.io/centos/centos:stream9 is not a bootc image]
============================================================= 2 passed in 5.68s =============================================================
```

**Comment by mvo5** [comment]
Thank you for the updated PR!

> @mvo5 I incorporated your patches. My editor applied my default formatting, not sure there is a formatter the project uses.

The formatitng changes are something we would like to avoid. The diff should just contain the functional changes. It might be easiest to just "git am" the two patches from https://github.com/osbuild/bootc-image-builder/pull/475#discussion_r1644068730 and force push. Note that the "happy" path of  ValidateHasContainerTags is tested implicitly by any of the "test_{build,manifest}.py" that test a valid image (e.g. test_manifest..yy::test_manifest_smoke).
 
> I had an issue with `test_manifest.py`. 5 of the tests were failing due to arch issues. Maybe this is environmental on my end?  Example:
> 
> ```
> Trying to pull quay.io/centos-bootc/fedora-bootc:eln...
> Error: error creating build container: choosing an image from manifest list docker://quay.io/centos-bootc/fedora-bootc:eln: no image found in image index for architecture x86_64, variant "", OS linux
> FAILED
[..]

Thank you! I will try to reproduce, is there anything unusual about your setup that might help me to reproduce?

**Comment by achilleas-k** [comment]
I got a bit lost reviewing that last commit with all the formatting changes so I edited the commits a bit.  Broke out the octal literal format change into its own commit, reverted the formatting changes to `test_manifest.py` and fixed the import order in a separate commit.

**Review by achilleas-k — APPROVED** [review]
Thanks for the contribution.  This LGTM!

**Review by achilleas-k — DISMISSED** [review]

**Comment by achilleas-k** [comment]
Rebased and resolved conflict.

**Review by achilleas-k — APPROVED** [review]

**Review by mvo5 — APPROVED** [review]
Thank you!

**Review by achilleas-k — APPROVED** [review]

**Pull request merged**